### ISSUE TITLE
What is the MSRV?

### ISSUE THREAD

**Issue by atouchet** [rust-lang/hashbrown#571]
I was looking at this repo and noticed that the minimum Rust version tested in CI is 1.74, the minimum version listed in Cargo.toml is 1.65, and the minimum version listed in the Readme and changelog is 1.63. Is there a reason for these discrepancies?

**Comment by clarfonthey** [comment]
`Cargo.toml` and `CHANGELOG.md` conflicting is due to the fact that MSRV will be bumped on the next release, but said release hasn't been published yet.

As for CI, it appears that the issue is that the MSRV of the various dependencies, specifically `allocator-api2`, is actually higher than the MSRV needed for hashbrown to compile. I guess that perhaps this should be fixed; cc @Amanieu.

Will poke around to see if `Cargo` actually has any open issues for linting this, since it feels like there should be a warning for having an MSRV lower than your dependencies.

**Comment by cuviper** [comment]
I've added CI checks in #572.

**Comment by Amanieu** [comment]
We're likely to release 0.16 soon to deal with #570, so this is a good opportunity to significantly bump the MSRV. I don't have any preference for a particular version, but @cuviper might.

**Comment by clarfonthey** [comment]
I know you expressed that you were okay bumping up all the way to 1.80.0, and at least when I was working on refactoring the code, I noticed that `NullPtr::add` was stabilised in… 1.80.0. So, that's one note in favour of that.

**Comment by cuviper** [comment]
My MSRV preference is probably more conservative than most -- I usually advocate for at least a year, which right now would be Rust 1.73 (2023-10-05). A little tighter than that could be 1.75.0 (2023-12-28), which is significant as the current Rust toolchain in RHEL 8 and 9, and also Ubuntu 20.04, 22.04, and 24.04. Debian stable still has 1.63, but they do have `rustc-web` now on 1.78.

In general, I don't think it should be bumped for no reason, and the example of `NonNull::add` seems like a weak motivation. Stuff like additional `const` power would be more motivating to me.

**Comment by clarfonthey** [comment]
Right, I mostly mentioned it to point out that there are things that could be directly useful even in the previous stable version and so keeping us back further will disrupt that. A lot of them in particular will directly affect the kinds of unsafe code that are used in the library and how we interact with them.

But obviously, yes, that one example could be easily to write manually, and it's not a massive notivating factor.

I think that setting one year back is a good starting point though, and we can talk more about it as things come up. We're not even using all the features the current, much older MSRV offers, so, that's worth fixing.

**Comment by cuviper** [comment]
As the `indexmap` maintainer -- it's a little unfortunate to have its MSRV forced up indirectly, but I also don't expect to keep supporting old versions forever. My usual rule of thumb is to support at least a year old, and even Rust 1.65 is now _two_ years old. When building for Debian 12, you can still lock it back to `indexmap v2.5.0` which uses `hashbrown v0.14`, as long as you refrain from using any new features.

(Please don't publish to crates.io with a max-limited dependency though!)

**Comment by s-arash** [comment]
FWIW, I've opened a similar issue on `allocator-api2`: https://github.com/zakarumych/allocator-api2/issues/20

**Comment by atouchet** [comment]
The Readme, Cargo.toml, and CI all now list the MSRV as 1.65.0. I think this has been sufficiently addressed.
